### PR TITLE
lib: prepare code for metarhia-common and metasync

### PR DESCRIPTION
**Pull request by tshemsedinov**
@aqrln This PR is not complete, just for your interest. I'll do it in multiple steps and I may need your help.

### CODE REVIEW

**Comment by tshemsedinov** [comment]
@aqrln done, but I rebased this branch with branch `style-and-opt` to avoid conflicts with eb7a704 `lib: refactor return style` so PR includes commit from different PR.

**Comment by tshemsedinov** [comment]
@aqrln rebased again

File: lib/remote-proxy.js [hunk 1]
```diff
@@ -55,15 +57,8 @@ RemoteProxy.wrapRemoteMethod = (instance, methodName) => {
...
-  let callback = args[args.length - 1];
-  args = Array.prototype.slice.call(args, 0, -1);
-
-  if (typeof(callback) !== 'function') {
-    args.push(callback);
-    callback = null;
-  }
-
-  this._connection.callMethod(this._interfaceName, methodName, args, callback);
+  const cb = common.cbExtract(args);
```
Comment (aqrln) [inline lib/remote-proxy.js]: Can we please call it `callback`? I'd prefer to use descriptive names when possible.

File: lib/socket.js [hunk 2]
```diff
@@ -83,12 +85,12 @@ class Transport extends EventEmitter {
 }
 
 const socketFactory = connect => (...options) => {
-  const callback = common.extractCallback(options);
+  const cb = common.cbExtract(options);
```
Comment (aqrln) [inline lib/socket.js]: Ditto.

File: lib/transport-common.js [hunk 3]
```diff
@@ -27,10 +29,10 @@ const common = require('./common');
 const newConnectFn = (
   connFactory, transportClass
 ) => (appName, client, ...options) => {
-  const callback = common.extractCallback(options);
+  const cb = common.cbExtract(options);
```
Comment (aqrln) [inline lib/transport-common.js]: Ditto.

File: lib/transport-common.js [hunk 4]
```diff
@@ -71,30 +73,21 @@ const newConnectFn = (
 const newConnectAndInspectFn = (connFactory, transportClass) => {
   const connect = newConnectFn(connFactory, transportClass);
   return (appName, client, interfaces, ...options) => {
-    const callback = common.extractCallback(options);
+    const cb = common.cbExtract(options);
```
Comment (aqrln) [inline lib/transport-common.js]: Ditto.

File: lib/transport-common.js [hunk 5]
```diff
@@ -71,30 +73,21 @@ const newConnectFn = (
...
       if (error) {
-        callback(error);
+        cb(error);
         return;
       }
-      const proxies = Object.create(null);
-      let errored = false;
-      const len = interfaces.length;
-      let count = 0;
+      const pCollector = metasync.collect(interfaces.length);
```
Comment (aqrln) [inline lib/transport-common.js]: Maybe something like `proxiesCollector` or even just `collector`. That lone "p" letter doesn't help readability much and looks like hungarian notation.

File: lib/ws-browser.js [hunk 6]
```diff
@@ -80,14 +82,15 @@ class Transport extends EventEmitter {
 //   appProvider - client application provider
 //
 const socketFactory = (url, callback) => {
+  const cb = common.cb(callback);
```
Comment (aqrln) [inline lib/ws-browser.js]: I'm not sure whether we need it here, it may be handled at a higher level.

/cc @lundibundi

File: lib/ws-internal.js [hunk 7]
```diff
@@ -124,13 +126,13 @@ const initServer = function(options, httpServer) {
 //             is established.
 //
 const connectionFactory = (webSocketConfig, ...options) => {
-  const callback = common.extractCallback(options);
+  const cb = common.cbExtract(options);
```
Comment (aqrln) [inline lib/ws-internal.js]: I'd prefer it to be `callback`.

File: package-lock.json [hunk 8]
```diff
@@ -45,12 +45,14 @@
       }
     },
     "ajv": {
-      "version": "4.11.8",
-      "resolved": "https://registry.npmjs.org/ajv/-/ajv-4.11.8.tgz",
-      "integrity": "sha1-gv+wKynmYq5TvcIK8VlHcGc5xTY=",
+      "version": "5.2.2",
```
Comment (aqrln) [inline package-lock.json]: There's a huge deal of unrelated changes and version bumps. Can you please revert all the changes to `package.json` and `package-lock.json` and run `npm i metasync metarhia-common` anew?

File: tools/common.js [hunk 9]
```diff
@@ -24,4 +13,18 @@ common.promisify = fn => (...args) => (
...
+  const exec = promisify(childProcess.exec);
+  return exec(cmd).then((stdout, stderr) => {
+    if (stderr) console.error(stderr);
+    return stdout;
+  });
+};
+
+const writeFile = promisify(fs.writeFile);
+
+module.exports = {
```
Comment (aqrln) [inline tools/common.js]: This change is unrelated. Can you please split it into another PR?

File: lib/ws-browser.js (hunk 6 above)
Comment (lundibundi) [inline lib/ws-browser.js]: This is also an internal function, there is always a callback here.

Comment (lundibundi) [inline lib/ws-browser.js]: This is also an internal function, there is always a callback here.

File: package-lock.json (hunk 8 above)
Comment (tshemsedinov) [inline package-lock.json]: @aqrln but this PR requires `metasync` and `metarhia-common` on `package.json`

Comment (aqrln) [inline package-lock.json]: Yes. But running `npm install metasync metarhia-common` will add them to `package.json` too.

**Comment by tshemsedinov** [comment]
@aqrln see dff1d345d2b4c341119919f2b7b75d1ccc6d5ec4 it is related to this PR because of naming but extracted to separate commit.

**Comment by tshemsedinov** [comment]
@aqrln rebased and ready

File: lib/cli/line-processor.js [hunk 10]
```diff
@@ -75,13 +75,13 @@ module.exports = class LineProcessor {
       callback(reportMissingArgument('Host'));
       return;
     }
-    const args = utils.split(tokens, ' ', 2);
-    let [scheme, authority] = utils.split(args[0], '://', 1, true);
+    const args = autocomplete.split(tokens, ' ', 2);
```
Comment (belochub) [inline lib/cli/line-processor.js]: `split` function here has nothing to do with autocomplete whatsoever, it is just used to split up the arguments provided by the user in CLI.

File: package.json [hunk 11]
```diff
@@ -37,6 +37,8 @@
   "dependencies": {
     "semver": "^5.4.1",
     "uuid": "^3.1.0",
+    "metarhia-common": "0.0.19",
```
Comment (nechaido) [inline package.json]: Shouldn't it be: `"metarhia-common": "^0.0.19",`?

File: lib/cli/autocomplete.js [hunk 12]
```diff
@@ -94,3 +91,11 @@ utils.iterativeCompletion = (inputs, depth, completer) => {
   }
   return [completions, help];
 };
+
+module.exports = {
+  filterKeys,
+  split,
```
Comment (nechaido) [inline lib/cli/autocomplete.js]: Why have you moved `rsplit` to `metarhia-common` but have not moved `split`?

File: lib/socket.js [hunk 13]
```diff
@@ -84,7 +85,7 @@ class Transport extends EventEmitter {
 
 const socketFactory = (connect) => {
   const resultConnectFn = (...options) => {
-    const callback = common.extractCallback(options);
+    const callback = common.cbExtract(options);
```
Comment (lundibundi) [inline lib/socket.js]: `extractCallback` and `cbExtract` are not equivalent, the latter has a lot of checks that are completely unnecessary here. I think it's better to place `extractCallback` in common and rename `cbExtract` to something that will represent its purpose, like `cbWrapExtract`.

File: lib/transport-common.js [hunk 14]
```diff
@@ -29,7 +31,7 @@ const common = require('./common');
 const newConnectFn = (
   connFactory, transportClass
 ) => (app, client, ...options) => {
-  const callback = common.extractCallback(options);
+  const callback = common.cbExtract(options);
```
Comment (lundibundi) [inline lib/transport-common.js]: Same as [cbExtract comment](https://github.com/metarhia/jstp/pull/246/files#r131100198).

File: lib/transport-common.js [hunk 15]
```diff
@@ -73,30 +75,21 @@ const newConnectFn = (
 const newConnectAndInspectFn = (connFactory, transportClass) => {
   const connect = newConnectFn(connFactory, transportClass);
   return (app, client, interfaces, ...options) => {
-    const callback = common.extractCallback(options);
+    const callback = common.cbExtract(options);
```
Comment (lundibundi) [inline lib/transport-common.js]: Same as [cbExtract comment](https://github.com/metarhia/jstp/pull/246/files#r131100198).

File: lib/ws-internal.js [hunk 16]
```diff
@@ -124,7 +125,7 @@ const initServer = function(options, httpServer) {
 //             is established.
 //
 const connectionFactory = (webSocketConfig, ...options) => {
-  const callback = common.extractCallback(options);
+  const callback = common.cbExtract(options);
```
Comment (lundibundi) [inline lib/ws-internal.js]: Same as [cbExtract comment](https://github.com/metarhia/jstp/pull/246/files#r131100198).

File: package.json (hunk 11 above)
Comment (aqrln) [inline package.json]: @nechaido both `0.0.19` and `^0.0.19` mean the same thing. Caret ranges only allow updates that don't modify the leftmost non-zero part of the version.

File: lib/applications.js [hunk 17]
```diff
@@ -17,7 +18,7 @@ const errors = require('./errors');
 //
 class Application {
   constructor(name, api, version) {
-    [this.name, this.version] = common.rsplit(name, '@');
+    [this.name, this.version] = common.rsection(name, '@');
```
Comment (lundibundi) [inline lib/applications.js]: I think that it's better to update `rsplit` function to support cases from [python](https://docs.python.org/3/library/stdtypes.html#str.rsplit) where I actually took it from and name it as such.

File: lib/socket.js (hunk 13 above)
Comment (tshemsedinov) [inline lib/socket.js]: Agree, in `metarhia-common` v0.0.21 we have `safeCallback` (renamed `cbExtract`) and `unsafeCallback` (renamed `cbUnsafe`). I placed your code of `extractCallback` into `common.unsafeCallback` because it's optimal implementation. Also I renamed `common.cb` to `once` because `once` will be clear and native for js world.

File: lib/applications.js (hunk 17 above)
Comment (tshemsedinov) [inline lib/applications.js]: I agree, we need full `rsplit` so I added issue here: https://github.com/metarhia/common/issues/91 but I think in this place `rsection` is ok because it is simpler and in multiple other cases `rsection` will be optimal.

File: lib/cli/autocomplete.js (hunk 12 above)
Comment (tshemsedinov) [inline lib/cli/autocomplete.js]: I changed `rsplit` to `common.rsection` because it is not compatible with `String.prototype.split()` This `split` is also not compatible with `String.prototype.split()` so I think we will either rename it or use `String.prototype.split()` instead.

File: lib/cli/line-processor.js (hunk 10 above)
Comment (tshemsedinov) [inline lib/cli/line-processor.js]: I agree, but see https://github.com/metarhia/jstp/pull/246/files#r132857348 We must arrange such common utilities to have code consistency over project, Meterhia stack and JS stories in general.

File: lib/transport-common.js (hunk 14 above)
Comment (tshemsedinov) [inline lib/transport-common.js]: https://github.com/metarhia/jstp/pull/246/files#r132848995

File: lib/transport-common.js (hunk 15 above)
Comment (tshemsedinov) [inline lib/transport-common.js]: https://github.com/metarhia/jstp/pull/246/files#r132848995

File: lib/ws-internal.js (hunk 16 above)
Comment (tshemsedinov) [inline lib/ws-internal.js]: https://github.com/metarhia/jstp/pull/246/files#r132848995

File: lib/transport-common.js (hunk 14 above)
Comment (tshemsedinov) [inline lib/transport-common.js]: In this case user can call it without callback so we need `common.safeCallback`

File: lib/socket.js (hunk 13 above)
Comment (tshemsedinov) [inline lib/socket.js]: In this case user can call it without callback so we need `common.safeCallback`

File: lib/transport-common.js (hunk 15 above)
Comment (tshemsedinov) [inline lib/transport-common.js]: In this case user can call it without callback so we need `common.safeCallback`

File: lib/ws-internal.js (hunk 16 above)
Comment (tshemsedinov) [inline lib/ws-internal.js]: In this case user can call it without callback so we need `common.safeCallback`

**Comment by tshemsedinov** [comment]
In most cases we use `extractCallback` (new name `unnsafeCallback`) to get callback from user-defined connection function, so user can pass `callback` optionally. If no callback passed exception will throw. In such cases we may choose one of the following ways:
- Use `unsafeCallback` and declare callback as required, so user will be responsible for inappropriate use.
- Implement `requiredCallback` that requires last argument to be callback and throw exception otherwise.
- Use `safeCallback` and have no problems but overhead.
- Change `safeCallback` behavior: use empty function but remove prevention of multiple calls, so it will be safe enough for above this cases. For more strict cases we can implement `onceCallback`.

I'd prefer to have in `metarhia-common` all those cases:
- `unsafeCallback(array): function or null`
- `safeCallback(array): function or emptiness`
- `requiredCallback(array): function or throw`
- `onceCallback(array): once(function) or emptiness`

@metarhia/jstp-core

File: lib/socket.js (hunk 13 above)
Comment (lundibundi) [inline lib/socket.js]: This is an internal function that always has callback, the user cannot call it.

File: lib/transport-common.js (hunk 14 above)
Comment (lundibundi) [inline lib/transport-common.js]: In this case not providing a callback is both an error and strange (you are calling a function to create a connection and not expecting a result), that's why I used extract here.

File: lib/transport-common.js (hunk 15 above)
Comment (lundibundi) [inline lib/transport-common.js]: Same as https://github.com/metarhia/jstp/pull/246#discussion_r132892967.

File: lib/ws-internal.js (hunk 16 above)
Comment (lundibundi) [inline lib/ws-internal.js]: Same as https://github.com/metarhia/jstp/pull/246#discussion_r132892234.

File: lib/socket.js (hunk 13 above)
Comment (tshemsedinov) [inline lib/socket.js]: This callback comes from `jstp.net.connect` and `jstp.net.connectAndInspect` and tls/ws/wss analogs. Try to remove this callack in `test/node/transport-net-tcp.js` for example.

Comment (nechaido) [inline lib/socket.js]: At the time it throws a TypeError:
```
> jstp.net.connect('name', null, 123);
undefined
> TypeError: callback is not a function
    at connFactory (/home/nechaido/Projects/metarhia/jstp/lib/transport-common.js:35:7)
    at Socket.onceErrorListener (/home/nechaido/Projects/metarhia/jstp/lib/socket.js:91:9)
    at Object.onceWrapper (events.js:316:30)
    at emitOne (events.js:115:13)
    at Socket.emit (events.js:210:7)
    at emitErrorNT (internal/streams/destroy.js:64:8)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickDomainCallback (internal/process/next_tick.js:218:9)
```
I am ok with it. `connect`is not a function that can be called without callback.

Comment (tshemsedinov) [inline lib/socket.js]: We need to avoid throwing, it terminates process, it's better to bubble an error to upper level without process termination.

Comment (tshemsedinov) [inline lib/socket.js]: But I can agree that `connect` is client-side function, client-side code is less durable in termination/restarting meaning.

Comment (belochub) [inline lib/socket.js]: @tshemsedinov, but it's a completely natural behavior in Node.js for `TypeError` to occur whenever you provide invalid arguments to the function (which is the case with providing no callback to the `connect` function).
This means that the user gets `TypeError` thrown only when he/she is using the library incorrectly.

**Comment by belochub** [comment]
Closing this PR, since it's become outdated to the point where it's easier to rewrite it from scratch compared to resolving all of the merge conflicts when rebasing on `master`.
I will also create a corresponding issue to keep as a reminder (as proposed by @tshemsedinov).
